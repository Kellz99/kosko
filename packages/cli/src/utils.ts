export function excludeFalsyInArray<T>(input: (T | undefined | null)[]): T[] {
  return input.filter(Boolean) as T[];
}
