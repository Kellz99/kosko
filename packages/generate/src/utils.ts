export function validateConcurrency(value = 10): number {
  if (value < 1) {
    throw new Error("Concurrency must be greater than 0");
  }

  return value;
}
