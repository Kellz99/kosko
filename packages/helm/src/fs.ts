import { getErrorCode } from "@kosko/common-utils";
import { cp, rename, rm, stat } from "node:fs/promises";

export async function move(src: string, dest: string): Promise<void> {
  try {
    await rename(src, dest);
  } catch (err) {
    if (getErrorCode(err) !== "EXDEV") throw err;

    await cp(src, dest, {
      recursive: true,
      errorOnExist: true,
      preserveTimestamps: true
    });
    await rm(src, { recursive: true, force: true });
  }
}

export async function fileExists(path: string): Promise<boolean> {
  try {
    const stats = await stat(path);
    return stats.isFile();
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;
    return false;
  }
}
