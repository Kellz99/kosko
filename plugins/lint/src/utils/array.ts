export function difference<T>(a: T[], b: T[]): T[] {
  const set = new Set(b);
  return a.filter((x) => !set.has(x));
}
