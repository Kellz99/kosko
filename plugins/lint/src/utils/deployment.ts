import type { IPodTemplateSpec } from "kubernetes-models/v1/PodTemplateSpec";
import type { ResourcePath } from "./manifest";
import type { PartialDeep } from "type-fest";
import { apiVersionToGroup, isRecord } from "@kosko/common-utils";
import type { ILabelSelector } from "@kubernetes-models/apimachinery/apis/meta/v1/LabelSelector";
import { getObjectValue } from "./object";

const deploymentLikeResources: ResourcePath[] = [
  {
    apiGroup: "",
    kind: "ReplicationController",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "Deployment",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "StatefulSet",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "DaemonSet",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "ReplicaSet",
    keys: ["spec"]
  },
  {
    apiGroup: "batch",
    kind: "Job",
    keys: ["spec"]
  },
  {
    apiGroup: "batch",
    kind: "CronJob",
    keys: ["spec", "jobTemplate", "spec"]
  },
  {
    apiGroup: "argoproj.io",
    kind: "Rollout",
    keys: ["spec"]
  },
  {
    apiGroup: "serving.knative.dev",
    kind: "Service",
    keys: ["spec"]
  },
  {
    apiGroup: "serving.knative.dev",
    kind: "Configuration",
    keys: ["spec"]
  }
];

export interface DeploymentLikeSpec {
  template?: PartialDeep<IPodTemplateSpec>;
  selector?: ILabelSelector;
}

export function getDeploymentLikeSpec(
  value: unknown
): DeploymentLikeSpec | undefined {
  if (!isRecord(value)) return;

  for (const resource of deploymentLikeResources) {
    if (
      typeof value.apiVersion === "string" &&
      apiVersionToGroup(value.apiVersion) === resource.apiGroup &&
      value.kind === resource.kind
    ) {
      const spec = getObjectValue(value, resource.keys);

      if (isRecord(spec)) {
        return spec;
      }
    }
  }
}
