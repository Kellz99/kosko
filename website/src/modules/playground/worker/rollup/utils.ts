export function getModuleURLForCDN(module: string) {
  return `https://cdn.skypack.dev/${module}`;
}
