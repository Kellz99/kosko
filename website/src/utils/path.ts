export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
