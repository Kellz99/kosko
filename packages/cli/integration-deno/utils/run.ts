import execa from "execa";
import { join } from "node:path";
import { projectRoot } from "@kosko/test-utils";

export async function runDenoCLI(
  args: string[],
  options: execa.Options = {}
): Promise<execa.ExecaReturnValue> {
  return execa(
    "deno",
    [
      "run",
      "-A",
      "--import-map",
      join(projectRoot, "packages/cli/templates/deno/import_map.json"),
      "npm:kosko",
      ...args
    ],
    {
      ...options,
      env: {
        LC_ALL: "en_US",
        ...options.env
      }
    }
  );
}
