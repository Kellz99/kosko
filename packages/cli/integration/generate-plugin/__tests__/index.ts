import { runNodeCLI } from "../../utils/run";
import { dirname, join } from "node:path";
import { mkdir, rm, symlink } from "node:fs/promises";

const testDir = dirname(__dirname);

async function linkModule(name: string, path: string) {
  const src = join(testDir, "modules", path);
  const dst = join(testDir, "node_modules", name);

  await mkdir(dirname(dst), { recursive: true });
  await symlink(src, dst, "dir");
}

afterEach(async () => {
  await rm(join(testDir, "node_modules"), { recursive: true, force: true });
});

describe("when plugin is a CJS package", () => {
  beforeEach(async () => {
    await linkModule("test-plugin", "cjs");
  });

  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-package.toml"],
      { cwd: testDir }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when plugin is a CJS file", () => {
  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-file-cjs.toml"],
      {
        cwd: testDir
      }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when plugin is a ESM package", () => {
  beforeEach(async () => {
    await linkModule("test-plugin", "mjs");
  });

  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-package.toml"],
      { cwd: testDir }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when plugin is a ESM file", () => {
  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-file-mjs.toml"],
      { cwd: testDir }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when multiple plugins are specified", () => {
  beforeEach(async () => {
    await linkModule("test-plugin", "cjs");
  });

  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-multi.toml"],
      { cwd: testDir }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when plugin does not exist", () => {
  test("should throw an error", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-package.toml"],
      { cwd: testDir, reject: false }
    );

    expect(result.exitCode).toEqual(1);
    expect(result.stderr).toContain(
      `Failed to resolve path for plugin "test-plugin"`
    );
  });
});

describe("when plugin is a TS file", () => {
  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-file-ts-cjs.toml"],
      {
        cwd: testDir,
        env: { TS_NODE_PROJECT: join(testDir, "modules/ts-cjs/tsconfig.json") }
      }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when plugin is a directory", () => {
  test("should transform manifests", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-dir.toml"],
      { cwd: testDir }
    );

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when transformManifest throws an error", () => {
  beforeEach(async () => {
    await linkModule("test-plugin", "transform-error");
  });

  test("should throw an error", async () => {
    const result = await runNodeCLI(
      ["generate", "--config", "./kosko-package.toml"],
      { cwd: testDir, reject: false }
    );

    expect(result.exitCode).toEqual(1);
    expect(result.stderr).toContain(`An error occurred in transform function`);
    expect(result.stderr).toContain(`Error: oops`);
  });
});
