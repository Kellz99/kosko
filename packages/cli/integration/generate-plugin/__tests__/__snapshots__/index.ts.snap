// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when multiple plugins are specified should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  labels:
    a: '1'
  annotations:
    b: '2'"
`;

exports[`when plugin is a CJS file should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when plugin is a CJS package should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when plugin is a ESM file should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when plugin is a ESM package should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when plugin is a TS file should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when plugin is a directory should transform manifests 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: foo"
`;

exports[`when validateAllManifests reports a error should throw an error 1`] = `
"
[4mcomponents/pod.js[24m - v1/ConfigMap test

[31mâœ– Test error message[39m

[31merror[39m - Found 1 error in total
[31merror[39m - Generate failed"
`;

exports[`when validateAllManifests reports a warning should not throw an error 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test"
`;

exports[`when validateAllManifests reports a warning should not throw an error 2`] = `
"
[4mcomponents/pod.js[24m - v1/ConfigMap test

[33mâš  Test warning message[39m

[33mwarn [39m - Found 1 warning in total"
`;

exports[`when validateManifest reports a error should throw an error 1`] = `
"
[4mcomponents/pod.js[24m - v1/ConfigMap test

[31mâœ– Test error message[39m

[31merror[39m - Found 1 error in total
[31merror[39m - Generate failed"
`;

exports[`when validateManifest reports a warning should not throw an error 1`] = `
"apiVersion: v1
kind: ConfigMap
metadata:
  name: test"
`;

exports[`when validateManifest reports a warning should not throw an error 2`] = `
"
[4mcomponents/pod.js[24m - v1/ConfigMap test

[33mâš  Test warning message[39m

[33mwarn [39m - Found 1 warning in total"
`;

exports[`when validateManifest reports multiple issues should throw an error 1`] = `
"
[4mcomponents/pod.js[24m - v1/ConfigMap test

[33mâš  Test warning 1[39m

[31mâœ– Test error 1[39m

[33mâš  Test warning 2[39m

[31merror[39m - Found 1 error and 2 warnings in total
[31merror[39m - Generate failed"
`;
